[2단계 - 지하철 노선도 미션] 동동(김동희) 미션 제출합니다.

bigsaigon333 opened the pull request:
# [2단계 - 지하철 노선도 미션] 동동(김동희) 미션 제출합니다.

@HyeonaKwon 님 안녕하세요 동동🏋🏼 입니다

지하철노선도 미션 2단계 PR 제출하오니 리뷰 부탁드립니다! 



> status 에 관한것을 훅으로 만들어보는 것도 좋을 것 같아요! https://github.com/woowacourse/react-subway-map/pull/16#discussion_r644089984

각 slice마다 status, message, reset, fetch 함수가 제각각이라 한번에 추상화된 훅으로 빼네는게 쉽지 않았습니다..

고민 결과, 공통된 로직을 추상화한 최상위 hook(`useStatus`)과,  각 slice별로 구체화한 hook(`useStationsStatus`, `useLinesStatus`, `useOverviewStatus`)으로 나누어서 추상화해보았습니다! (뭔가 2% 부족한 듯한 느낌이 듭니다만... :sob: )



## 1.  [:fire: 데모사이트 :fire: ](https://bigsaigon333.github.io/react-subway-map)

## 2. [:airplane: VSCODE  in web :airplane: ](https://github1s.com/bigsaigon333/react-subway-map/tree/step2)

## 3. 구현한 기능 목록 - 2단계
- [x] 전체 보기 기능

bigsaigon333 committed:
feat: 전체 보기 페이지를 추가한다

bigsaigon333 committed:
feat: 전체 보기 페이지 내 노선별 정보를 표시한다

bigsaigon333 committed:
refactor: RegExp 변수명을 수정한다

bigsaigon333 committed:
refactor: status 에 관한 훅을 추출한다

bigsaigon333 committed:
style: 불필요한 eslint 주석을 제거한다

bigsaigon333 committed:
feat: 전체 보기 화면에서 각 구간별 거리를 표시한다

woowapark reviewed (changes requested):
동동 안녕하세요! 현아님을 놓아드리면서 대신 리뷰를 남깁니다. 
세세한 부분 몇가지는 코드 라인에 조금 더 남겨둘게요!

### 👏👏👏
- 커스텀 훅으로 만들어서 재사용해보신 소감이 궁금하네요! 고민해서 적용해본 것에 먼저 박수! 👏👏👏
- 함수가 제각각이라 한번에 추상화하기 어려웠다고 남겨주셨는데, 공통 로직 단위로 각각을 비슷한 형식으로 정리해보고 -> 정리된 걸 다시 묶어서 추상화해보는 방법도 좋은 것 같아요. 
    - fetch하는 방식들만 최대한 일관되게 바꿔보고 -> reset하는 위치를 일관되게 바꿔보고 -> message 사용하는 위치도 같이 점검해본 다음 -> status 를 사용하는 형식까지 맞춰보고 
    - 요렇게 각 부분별로 정리한 상태에서, 그 중에서 반복되는 부분을 추상화해서 빼는 방식으로요. (물론 이미 이렇게 했을 수도 있고 이렇게 해도 쉽지는 않습니다만....그래도 조금이나마 단계적으로 접근하는 방법으로..! ㅋㅋ)

### 페이지마다 useStatus를 감싼 hook을 따로 만든 부분
- 우선 반복되는 로직에 해당하는 부분은 `useStatus` 내부에 이미 들어가 있고,     
  `useLinesStatus`, `useOverviewStatus`는 단지 selector만 추가하는 형태라 
    `useLinesStatus`, `useOverviewStatus`를 별도의 hook으로 다시 묶는 것의 효과가 크지는 않아 보여요. 🤔  
     재사용을 위해서라기에는 실제로 사용하는 곳도 한 군데씩인 것 같고요!
    - 페이지에서 이미 selector들도 쓰고 있으니 굳이 따로 묶을 필요 없이 페이지에서 바로 useSelector + useStatus를 사용하거나
    - `useStatus` 훅에서 아예 selector를 인자로 받는 방식도 고려해볼 수 있을 것 같습니다.
    
### 전체 구조상 일관성
- 위에서 말한 hook을 사용하는 부분과 함께, component와 page 각각의 역할과 구성 방식에 조금 더 일관성이 있으면 더 좋을 것 같습니다!

#### useStatus 훅을 사용하는 위치, 역할 - 데이터를 다루는 로직은 어느 영역에 위치해야 할까요?
- overview는 **페이지에서** `useOverviewStatus`를 사용하고, lines는 **컴포넌트** 단위에서 `useLinesStatus`를 실제로 사용하고 있는데요. 
  비슷한 구조에서 비슷한 역할을 하는 로직이니만큼 어느쪽으로든 일관된 형태로 사용하면 좋을 것 같아요
- 이 경우, 외부 데이터와 연동하는 로직을 어느 영역에 위치시킬 것인지에 대한 일관된 기준을 고민해보는 일이 필요할 것 같네요
    - 동동이 아닌 다른 개발자가 이 어플리케이션에 새로운 기능을 추가한다고 했을 때,     
        외부 API와 연동되는 로직을 어디에 추가할 지, store와 연결되는 부분은 어디에 위치시킬 지 를 쉽게 파악할 수 있을 지를 함께 고민해보면 좋을 것 같습니다 :) 

#### 컴포넌트 구성 방식
- `<Main>` 역시 어떤 경우에는 page단에서 직접 구성하고, 어떤 경우에는 아예 component/ 하위에 `<LinesMain>` 같이 따로 만들고 있는데요..! 
    - 어디에서 어떻게 구성할 지는 선택의 영역이지만, 선택한 방식은 일관되게 적용해두어야 이후 유지보수할 때 어떤 영역을 찾아서 수정해야 할 지 파악이 쉬울 것 같아요 😃 
    - 사실상 같은 역할을 하는 컴포넌트 로직인데 어떨 때에는 page 하위에 있고, 어떨 때에는 component 하위에 있다면 이후에 비슷한 컴포넌트를 추가하게 될 때 어떤 기준으로 어디에 추가할 지 헷갈릴 수 있지 않을까요 ?-?
- 비슷한 맥락에서(?) `LinesMain`는 컴포넌트가 -> 페이지에 있는 걸 참조하는 형태로 구성되어 있는데요. 
   - 지금같은 경우는 이 컴포넌트가 다른 페이지에서 쓰이지 않고 있어서 문제는 없겠지만, 
      컴포넌트가 페이지에 있는 걸 역으로 참조하고 있는 경우, 이 컴포넌트를 다른 페이지에서 쓰게 될 때 이슈가 될 수 있을 것 같아요

### UI 리뷰
- 로딩 Ui에서 화면 전체에 딤드 레이어를 까는 것 때문에 네트워크가 빠를 때 로딩 UI가 노출되는 시간이 짧으면 화면이 깜빡이는 듯한 현상이 있습니다. 가운데에 보이는 컴포넌트에서만 로딩이 일어나는 거니, 화면 전체에 딤드 레이어를 까는 것보다는 해당 컴포넌트 내부에서만 로딩 아이콘을 적용해보아도 좋을 것 같아요.
- 요청이 실패한 경우, alert이 뜨고 이전에 그려진 데이터는 그대로 남아있습니다. 요것이 의도한 렌더링 결과가 맞을까요?

woowapark commented on src/pages/Overview/index.jsx:
```diff
@@ -0,0 +1,36 @@
+
+  return (
+    <>
+      <Loading isLoading={status === STATUS.LOADING} bgOpacity="10" />
```
요건 그냥 개인 의견! 
모든 `<Loading />` 컴포넌트가 현재 `bgOpacity="10"` 옵션으로 쓰이고 있는 것으로 보이는데요, 
요게 기본 옵션이라고 하면 `<Loading />` 컴포넌트 내부에서 기본값을 지정해두고, 다른 옵션값을 사용할 때에만 추가로 지정해서 쓸 수 있게 해준다면 이 컴포넌트를 사용하는 쪽에서 신경써야 할 부분이 하나 줄어들지 않을까 합니다 😃

bigsaigon333 replied:
1. 회원가입 페이지와 로그인 페이지를 먼저 구현하고, 역 관리, 노선 관리, 구간 관리를 구현하였습니다. 

2. 회원가입 페이지와 로그인 페이지 모두  ` <Loading />` 을 사용하고 있는데, 이 때 둘다 `bgOpacity`를 40으로 사용하여, 기본값을 40으로 하였습니다. 
```jsx
Loading.defaultProps = {
  bgOpacity: "40",
};
```
3. 역 관리, 노선 관리, 구간 관리에서는 `bgOpacity` 40은 너무 진한듯하여, 10으로 하여 사용하였습니다.




`구간 검색` 기능이 더 추가된다고 한다면, `<Loading />` 의 bgOpacity`는 10으로 아마 할 것 같습니다. 

유지보수에 효율적이려면 변경이 더 잦은 코드의 복잡도를 낮춰야 할 것 같은데, 

거의 변경할 여지가 없는 회원가입 페이지와 로그인 페이지에서 쓰이는 bgOpacity의 값인 40으로 기본값을 해둔게 

조금 아쉬운 선택인 것 같습니다... :crying_cat_face:

woowapark replied:
아하 사용하는 쪽을 찾아보다가 이미 기본값 주신 게 있었던 걸 놓쳤네요 😅      
동동이 이미 잘 말씀해주신 것처럼 변경이 더 잦은 쪽에서 신경 써야 하는 걸 줄일 수 있으면 좋을 것 같아요! 👍
덧붙이자면 디자인 가이드가 있거나 디자인을 함께 하고 있는 상황에서는     
디자인에서 '기본'으로 취급하는 건 어떤 상태인지도 확인해 서로 생각하는 '기본'을 일치시킬 수 있는 방향을 고민해볼 수도 있을 것 같구요 :)

bigsaigon333 replied:
`Loading` 컴포넌트의 Dimmed 된 영역의 생상을 변경시키지 않는 것으로 수정하였습니다.. 이에 위에 말씀드렸던 사항은 현재버젼 코드에서는 해당사항이 없습니다..! :sweat_smile:

woowapark commented on src/components/OverviewListItemDetail/index.jsx:
```diff
@@ -0,0 +1,39 @@
+import PropTypes from "prop-types";
+import cx from "classnames";
+
+const formatDistance = (dist) => dist.toLocaleString("en-US");
```
formatting같은 경우, 컴포넌트의 다른 로직과 상관없이 범용적으로 사용할 수 있는 경향이 크니 반복해서 사용하게 된다면 별도의 util로 분리해도 좋을 것 같습니당~

bigsaigon333 replied:
`formatDistance`가 지엽적으로만 사용되고 있다고 생각해서 별도로 분리를 안했었는데, 확인해보니 2군데에서 사용되고 있네요 :open_mouth: 

`utils/format.js` 로 분리하였습니다 :+1:

woowapark commented on src/components/LinesMain/index.jsx:
```diff
@@ -1,42 +1,23 @@
 import Main from "../@shared/Main";
 import Button from "../@shared/Button";
 import LinesList from "../LinesList";
+import { useLinesStatus } from "../../pages/Lines/hooks";
```
(요것도 개인 취향일 수는 있습니다만..ㅋㅋ)
비슷한 import문끼리 모아서 import 구문 영역도 잘 정리해주면 좀 더 깔끔쓰하게 읽힐 것 같아요우

bigsaigon333 replied:
앗 대부분 비슷한 import문끼리 모아놨다고 생각했는데, 여기는 놓쳤네요 ㅠㅠ

성격이 비슷한 import문끼리 모으면서도 path가 가까운걸 아래에 쓰고 있었는데, 아래의 순서대로 import문을 정리하였습니다. 
path가 뒤죽박죽이 되버리니 조금 어색한 느낌이 들긴 하는데, 확실히 어떤 hooks를 쓰는지 한눈에 보여서 그건 좋은것 같습니다!
다음에는 hooks는 hooks 디렉토리에, slice는 slice 디렉토리에 담았다면 더 깔끔해질 수 있지 않았을까... 생각해봅니다 :smile: 

- packages
- hooks
- constants
- utils
- actionCreators
- components
- assets

woowapark replied:
요런 건 특히나 정답이 있는 건 아니니까요!      
역할/의미 구분이 명확하게 드러나는지 & 스스로 + 협업하는 동료가 파악하기에도 좋을 지만 꾸준히 고민해보면서 구성하면 될 것 같습니다~🙂

woowapark commented on src/components/LinesForm/validator.js:
```diff
@@ -1,11 +1,11 @@
 export const isValidLineName = (name) => {
-  const lineNamePattern = /^[가-힣0-9]{2,10}$/;
+  const lineNameRegex = /^[가-힣0-9]{2,10}$/;
```
validation 로직은 가장 쉽게 단위 테스트를 쓸 수 있고 + 가장 빨리 효과를 볼 수 있는 부분일 것 같아요      
'잘못된 값'들에 대한 테스트 케이스들을 써두면 직접 테스트해야 하는 케이스를 줄이고 이후 validation 로직을 수정해야 하는 경우에도 조금 더 안정적으로 수정이 가능하니 이후에 고려해보셔요~

bigsaigon333 replied:
흑흑 validation 테스트는 이번 level 3 때 꼭 적용해보겠습니다!!

woowapark commented on src/components/OverviewListItem/index.jsx:
```diff
@@ -0,0 +1,48 @@
+    sections: OverviewListItemDetail.propTypes.sections,
+    color: PropTypes.string.isRequired,
+    distance: PropTypes.number.isRequired,
+  }).isRequired,
```
꼼꼼한 propTypes 👍

woowapark commented on src/pages/Overview/index.jsx:
```diff
@@ -0,0 +1,36 @@
+  return (
+    <>
+      <Loading isLoading={status === STATUS.LOADING} bgOpacity="10" />
+      <Main>
```
(덧붙여, `Main` 같은 경우는 아예 html 엘리먼트로도 있는 예약어다보니 컴포넌트의 이름으로 쓰기에는 조금 혼동을 일으킬 수 있을 것 같아요)

bigsaigon333 replied:
앗 사실 일부러 그부분을 의도한 네이밍인건데요.. 

html element인 `<main />` 을 쓰고 싶은 경우에는 `<Main />` 을 쓰라는 의미를 다른 개발자에게 전달하고 싶었습니다.

margin 및 flexbox를 이용한 가운데 정렬, 최소너비, 높이 등을 `<Main />`에서 fix해두고, 이 어플에서는 일관되게 적용하고 싶었습니다. 

그런 의도로 작성한 컴포넌트 치고는 말랑말랑(?)하지 못한 걸까요... ? :crying_cat_face:

bigsaigon333 committed:
refactor: 거리 표시 형식을 지정하는 유틸함수를 분리한다

bigsaigon333 committed:
style: 모듈 import 순서를 정렬한다

1. packages
2. hooks
3. constants
4. utils
5. actionCreators
6. components
7. assets

bigsaigon333 committed:
fix: 로딩 컴포넌트가 부모 컴포넌트 내에서만 표시되도록 한다

bigsaigon333 committed:
feat: Alert 컴포넌트를 구현한다

bigsaigon333 committed:
feat: useBoolean 훅을 구현한다

bigsaigon333 committed:
fix: 모달 노드가 매 렌더링마다 추가되고 제거되던 오류를 수정한다

bigsaigon333 committed:
refactor: 노선 관련 비동기 로직을 노선 관리 페이지에서 제어한다

woowapark commented:
@bigsaigon333 
동동 혹시 추가로 수정하실 사항이 있을까요?-?    
수정하신 뒤로 추가 리뷰 요청 알림은 받지 못한 것 같아서요..!

bigsaigon333 committed:
refactor: 구간 관련 비동기 로직을 구간 관리 페이지에서 제어한다

bigsaigon333 committed:
refactor: 노선 정보 조회 api콜을 useFetchOnce 훅으로 수행한다

bigsaigon333 committed:
refactor: 전체 보기 페이지의 컴포넌트 구성을 노선 관리, 구간 관리와 일치시킨다

bigsaigon333 committed:
refactor: 역 관리 페이지에서 StationsMain 컴포넌트를 분리한다

bigsaigon333 committed:
feat: Alert가 표시될 때 확인 버튼에 focus가 가도록 한다

bigsaigon333 committed:
refactor: 로그인 페이지에서 LoginMain을 분리한다

bigsaigon333 committed:
refactor: 회원가입 페이지에서 SignupMain을 분리한다

bigsaigon333 committed:
refactor: 서버 선택 페이지의 불필요한 api콜을 제거한다

bigsaigon333 committed:
fix: 사용되지 않는 커스텀 훅을 제거한다

bigsaigon333 committed:
refactor: Alert 컴포넌트를 추상화한다

bigsaigon333 committed:
refactor: 비정상적인 접근에 대한 알림에 Alert컴포넌트를 이용한다

bigsaigon333 committed:
feat: Confirm 컴포넌트를 구현하여 window.alert를 대체한다

bigsaigon333 committed:
fix: 역 관리 reset action 실행시 메세지도 초기화한다

bigsaigon333 committed:
feat: 회원가입, 로그인 페이지 접속시 처음 Input을 포커스한다

bigsaigon333 commented:
> @bigsaigon333
> 동동 혹시 추가로 수정하실 사항이 있을까요?-?
> 수정하신 뒤로 추가 리뷰 요청 알림은 받지 못한 것 같아서요..!

@woowapark  헉 빨리 리뷰 재요청을 드렸어야 했는데 늦어서 죄송합니다.. :sob:  지금 이제 커밋은 다 마무리했고, 코멘트 추가로 달고 있습니다..! 
금방 완료한 후 리뷰 재요청 버튼 누르겠습니다...!

bigsaigon333 commented:
> 커스텀 훅으로 만들어서 재사용해보신 소감이 궁금하네요! 

1. 훅 자체가 익숙하지 않은 상황에서 커스텀 훅을 조금씩 써보는게 처음에 쉽지는 않았습니다. 
조금 익숙해지고 난 뒤에 다시 생각해보니, 훅은 React에서 제공하는 기본 api 라고 할때, 
커스텀 훅은 그런 기본 api를 묶어서 추상화하는 것이므로, 반복되는 코드를 추상화해서 별도의 유틸함수로 만드는 것이랑 
별반 다를바가 없구나.. 라는 생각이 들었습니다. 
2. 혼자서 어떻게 하면 좋을지 고민하는 것도 좋지만, 다른 코드를 참고해보는 것도 좋다는 생각도 들었습니다. 
`react-use` 를 보면서 아 사람들은 이렇게 커스텀 훅을 짜는구나 라는 감을 익혔습니다. 
3. 이번 공원의 리뷰를 반영하면서 일관되게 느낀 점인데, 제가 짠 코드가 컴포넌트-컴포넌트 간 결합도도 높고 훅과 컴포넌트의 결합도도 상당히 높았구나 라는걸 깨달았습니다. 컴포넌트를 말랑말랑(?)하게 짜라는 얘기는 진짜 많이 들었던 것 같은데, 훅도 말랑말랑하게 짜야 재사용도가 높아진다는 걸 깨달았습니다. 

> 전체 구조상 일관성

각 페이지마다 구조가 정말 들쑥날쑥이었습니다... 
각 page의 컴포넌트 구성을 아래와 같이 일치시켰습니다. 
```jsx
const Page = () => {
  // 마운트시 fetch
  
  return (
  <>
    <Loading />
    <Alert />
    <Main />
    <Modal />
  </>
}
```
- 비동기 호출에 따른 status 관리를 모두 Page에서 수행하도록 하였습니다. 
`Alert` Component를 구현하여 status 변화에 따른 message를 표시하게끔 수정하여 `window.alert`를 모두 날려버렸습니다. (UI부분을 `window.alert` 같은 사이드 이펙트를 이용해서 쉽게 해결하려고 하면 더 고생한다는걸 뼈저리게 느꼈습니다..)

> UI 리뷰
1. Loading 컴포넌트를 수정하여 부모 컴포넌트의 정 가운데에 표시되도록 하였습니다. 기존에는 무조건 스크린의 정가운데에 표시되어 Loading 컴포넌트의 자유도를 떨어트렸습니다. 또한, Dimmed 부분의 색상도 제거하여, 눈의 피로도를 줄였습니다(?).
2. 요청이 실패한 경우에 기존에 유저가 입력한 값이 그대로 남아 있는 것은 제가 의도한 부분입니다. 예를 들어 회원가입 화면에서 유저가 폼을 제출하고 나서 ID가 중복되었다는 결과를 받은 경우에, 기존의 값이 그대로 남아있어야 틀린 부분(ID)만 수정할 수 있기 때문입니다. ..! 또한 `window.alert`를 모두 제거하고 기존의 렌더링 된 화면 위에 `Alert` component가 표시되게끔 모두 수정하였습니다.

bigsaigon333 commented:
@woowapark  리뷰를 2주도 전에 해주셨는데 이제서야 리뷰 반영을 해서 죄송합니다 ㅠㅠ 

구조 전체에 대한 고민을 하면서 기존에 짰던 코드를 전체 다 볼 수 있어서 좋았습니다 ㅎㅎ..
(과거의 나 왜 이렇게 코드 작성했니...)

코멘트 및 리뷰 반영한 사항 확인 부탁드립니다~ :bowing_man: :bowing_man: :bowing_man: :bowing_man:

woowapark reviewed (approved):
와 동동 엄청 꼼꼼히 수정해주셨네요...!     
커밋 단위로 피드백 반영해주신 부분 잘 확인하였습니다 :)      
> 부모 컴포넌트의 정 가운데에       

요런 것도 재사용성이 훨씬 개선되었을 것 같아서 좋네요!    

> react-use 를 보면서   

무조건 따라하면 안되겠지만, 기본적으로 어느정도 사용이 많이 되고 있는 오픈소스 라이브러리에서 어떻게 하고 있는 지를 보는 것도 좋은 학습 방법이라고 생각해요 :) 물론 '모르겠지만 여기에서 이렇게 하는 걸 보니 좋은 방식이겠지~' 라는 생각으로 도입하면 위험하겠죠?! 🙅‍♂️      

`$modal`     
요런 건 jquery쪽에서 시작된- 엘리먼트를 구분하기 위한 컨벤션으로 알고 있는데 🤔       
컨벤션을 사용한 의도도 같을까용?-?      
다른 컨벤션이 있는지는 제가 모르지만 🤔 만약 jquery쪽에서 나온 컨벤션이 맞다면 헝가리안 표기법같이 어느정도는 굳이 새로 쓰는 코드에 적용하지는 않는 표기법일 수 있을 것 같아요.      
한번 확인해보시기를 추천드립니다~  

> 요청 실패 후 UI가 그대로 있는 부분         
아하 의도하신 부분이 어떤 것인지 이해하였습니당     
제가 확인했던 케이스는 전체보기 탭을 보고 -> 다른 탭 갔다가 네트워크 오류가 생긴 후에 -> 전체보기 탭을 다시 돌아왓을 때에도 이전에 성공한 결과로 렌더 된 게 남아있더라구요. 의도하신 흐름이 맞을까 궁금해서 코멘트 남겨보았었어요 :) 

전체적으로 구조상으로 일관성이 많이 정리된 것 같아요..!     
수정해보시면서 코멘트 드린 것 이상으로 고민도 많이 해주신 것 같구요 👍     
프로젝트 진행하면서도 일관성, 재사용성, 에러 처리, 테스트 등등 다양하게 적용해볼 수 있는 기회가 되셨으면 좋겠습니다 :)    
화이팅~!    

요 PR은 머지할게요!
